crates/pallet-messenger/src/lib.rs

```diff
@@ -0,0 +1,244 @@
+/// Nonce is always increasing.
+pub type Nonce = U256;
+
+/// Channel describes a bridge to exchange messages between two domains.
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo, MaxEncodedLen)]
+pub struct Channel {
+    /// Channel identifier.
+    pub(crate) channel_id: ChannelId,
+    /// State of the channel.
+    pub(crate) state: ChannelState,
+    /// Next inbox nonce.
+    pub(crate) next_inbox_nonce: Nonce,
+    /// Next outbox nonce.
+    pub(crate) next_outbox_nonce: Nonce,
+    /// Latest delivered outbox message nonce.
+    pub(crate) latest_delivered_message_nonce: Nonce,
+    /// Maximum outgoing non-delivered messages.
+    pub(crate) max_outgoing_messages: u64,
+}
+
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+pub struct InitiateChannelParams {
+    max_outgoing_messages: u64,
+}
+
+#[frame_support::pallet]
+mod pallet {
+    use crate::{Channel, ChannelId, ChannelState, InitiateChannelParams, U256};
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+    use sp_runtime::ArithmeticError;
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+        // Domain ID uniquely identifies a Domain
+        type DomainId: Parameter + Member + Default + Copy + MaxEncodedLen;
+    }
+
+    /// Pallet messenger used to communicate between domains and other blockchains.
+    #[pallet::pallet]
+    #[pallet::generate_store(pub(super) trait Store)]
+    pub struct Pallet<T>(_);
+
+    /// Stores the next channel id for a foreign domain.
+    #[pallet::storage]
+    #[pallet::getter(fn next_channel_id)]
+    pub(super) type NextChannelId<T: Config> =
+        StorageMap<_, Identity, T::DomainId, ChannelId, ValueQuery>;
+
+    /// Stores channel config between two domains.
+    /// Key points to the foreign domain wrt own domain's storage name space
+    #[pallet::storage]
+    #[pallet::getter(fn channels)]
+    pub(super) type Channels<T: Config> =
+        StorageDoubleMap<_, Identity, T::DomainId, Identity, ChannelId, Channel, OptionQuery>;
+
+    /// `pallet-messenger` events
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Emits when a channel between two domains in initiated.
+        ChannelInitiated {
+            /// Foreign domain id this channel connects to.
+            domain_id: T::DomainId,
+            /// Channel ID of the said channel.
+            channel_id: ChannelId,
+        },
+
+        /// Emits when a channel between two domains in closed.
+        ChannelClosed {
+            /// Foreign domain id this channel connects to.
+            domain_id: T::DomainId,
+            /// Channel ID of the said channel.
+            channel_id: ChannelId,
+        },
+
+        /// Emits when a channel between two domains in open.
+        ChannelOpen {
```

Comment:
Seem to be more consistent.

```suggestion
        ChannelOpened {
```

Reply:
hmmm I think `open` should work here very well since it presents the channel state unless you strongly feel about it

Reply:
The others use the verb in the past form, here uses the state after the operation was applied, that's why I feel it a bit inconsistent, no strong stand though, `open` is fine to me as well :D

Reply:
The others use the verb in the past form, here uses the state after the operation was applied, that's why I feel it a bit inconsistent, no strong stand though, `open` is fine to me as well :D